### ISSUE TITLE
Onshore transport - needed for autocyclicity

### ISSUE THREAD

**Issue by EmiliaJarochowska** [MindTheGap-ERC/CarboKitten.jl#73]
I should have identified this earlier, but it slipped my attention - sorry! For the emergence of autocyclicity in the absence of external drivers, it seems that onshore transport is required. Please see Cases 2 and 3 in [Xi & Burgess (2022)](https://onlinelibrary.wiley.com/doi/abs/10.1111/sed.12974). They write:

> As in previous models (Burgess et al., 2001), onshore prevailing winds and currents are assumed to produce a time-averaged net onshore advective sediment flux.

In top (plane) view, this is the regularity obtained:
![Untitled](https://github.com/user-attachments/assets/4daddb90-c245-4118-a1ad-2d8d6b7fc5a5)

Our version of CarboCAT doesn't have this feature, I wrote to Peter asking for that specific part of the code. This issue needs discussion between us first before anyone works on it.

**Comment by Burgesski** [comment]
I have had a look at the version of CarboCAT that Haiwei modified and used to produce these autocycle models. It's just a smidgeon convoluted ;-) so apologies for that, but I think actually it can be reasonably easily unravelled to understand the key elements in the sediment transport formulation that produces the shoreline/island progradation that generates the shallowing-upward autocycles.

The key functions are the various versions of the sediment transport calculation function and they seem to be calculateTransportStraight, calculateTransportPlatformTopShoaling, and calculateTransportPlatformTopShoalingWaveRefraction

I know, I know, multiple versions of essentially the same function calculate the sediment transport in slightly different ways is slightly sub-optimal coding practice ;-) but the good news is that the most important element in the calculation is the same in each function.

In each case a volume (thickness actually) of sediment is entrained, based on a parameter  indicating the proportion of produced sediment to be transported, as per standard CarboCAT practice. Looking in calculateTRansportStraight.m in lines 133 - 150 we can see that this sediment is then moved across the platform top following a defined vector, for example straight up the y-axis of the model grid, so long as the following while loop condition is met

while foundLocalHigh == false && endOfFlowVectorReached == false && flowLength > 1

flowLength and endOfFlowVectorReached just deal with no transport path defined and end of defined path reached, for example at the end of the grid, so the foundLocalHigh flag is the element we are most interested in, and this is defined via

if grad > glob.depoGradThreshold
                    foundLocalHigh = true;
                    DepobyLocalHighs(flowVect(2,j), flowVect(1,j))=1;
                end

which sets foundLocalHigh to true as the while loop terminating flag when the gradient into the next cell along the transport path exceeds a defined threshold.

So in summary, sediment is entrained from each producing point on the grid, and then transported across the grid until reaching a point where there is a topographic gradient to the next grid cell in the transport path, from deeper water into shallower water, that exceeds a defined threshold value, at which point the sediment is deposited.

If you can implement something like this in CarboKitten I would anticipate shoreline/island progradation behaviour should result, but I'm not sure how straight forward this will be to do given the diffusion approach adopted in CarboKitten? Perhaps we need to discuss this further??

cheers, P.

**Comment by HannoSpreeuw** [comment]
From a physics point of view, at a first glance the [Active Layer Transport model](https://mindthegap-erc.github.io/CarboKitten.jl/dev/active-layer-transport/) seems very well suited to model onshore transport, by making the advective velocity periodic, i.e. make it match the tidal cycles.

**Comment by EmiliaJarochowska** [comment]
I think we're now aiming for wave transport, so not periodic, just offshore. Tides would add more complexity to the problem.

**Comment by Burgesski** [comment]
Yes indeed, no need for it to be periodic tides, there just needs to be a long-tern net onshore flux with accretion of the sediment onto the "front" of the shoreline such that it can drive progradation, however is most appropriate to do that

**Comment by HannoSpreeuw** [comment]
But that sounds really simple. 

Can we get away with such a simple model, compared to much more sophisticated models,
such as [Edmonds & Slingerland (2010)](https://doi.org/10.1038/ngeo730)?

**Comment by HannoSpreeuw** [comment]
Allright, but pls let me be blunt and ask for more help:
How would one best model the waves within the Active Layer Transport model?

Any response is good, also blunt responses, such as: _Perhaps you should figure this out yourself_.

**Comment by Burgesski** [comment]
I don't think specific modelling of specific wave processes is really relevant. In the short term waves produce a longshore drift effect more than an onshore transport effect. We are trying to model a much longer-term geological process of cross-platform sediment flux, specifically onshore flux. That is likely to be a time-averaged result of several different processes, including but not limited to waves. So if it was me, I would just do it the way we did it in Xi and Burgess (2022) because the main aim here is to produce that triangle diagram, not really to produce a new model of wave and tide-related sediment transport etc. So, in summary, keep it simple - move sediment across the platform, accrete it to the shoreline to drive shoreline progradation. Do that in the simplest way possible. That would be my suggestion.

**Comment by HannoSpreeuw** [comment]
Thanks for that, I'll keep it simple.

So one way of proceeding would be adapting [Johan's proposal from a couple of months ago](https://github.com/MindTheGap-ERC/CarboKitten.jl/blob/73-onshore-transport/docs/src/onshore-transport.md) of doing nothing but adding a vector component $P_f w_f(\eta)$ to the flux:

$$q_f = -\nu_f P_f \nabla \eta + P_f w_f(\eta).$$

@jhidding has introduced a lateral advection $w_f$ here.

How do the others feel about this approach?

**Comment by EmiliaJarochowska** [comment]
I think this is quite close to what we intended. What we discussed in November are:
- should there be explicit accounting for the depth of wave base, so $w_f$ is bounded (zero above certain wd) because waves don't act at high depths (not all models will achieve such depths)
- we have to be very specific about ignoring longshore drift and why it does not affect our conclusions. Because what we are modeling is that a certain amount of sediment accretes onshore as a result of flux directed shoreward, whereas in reality it accretes by evening out the shoreline if it's uneven (erode what sticks out, accrete what is concave). As long as we assume the initial topography to be a ramp, it does not have any effect whatsoever, but one can assume any other topography and then the shoreline will be uneven. In such case, this simplest model will accrete even in the parts sticking out, which is the opposite to what happens in real life. We could ignore this for now and add complexity later.

@xyl96 is on holidays so he cannot participate in this discussion for another week, we have to move forward without him.

**Comment by jhidding** [comment]
C is easy, you should be able to understand it as well as Fortran ;)

**Comment by EmiliaJarochowska** [comment]
Fair, I looked at it but the barrier seems that I don't know the structure of the model, how is everything stored. I'd need to read the core of SedFlux. Sounds like on of our "what could possibly go wrong" undertakings 😂

**Comment by Burgesski** [comment]
Ok, if people are pulling things around with an immensely complicated set of ropes and rollers, reinventing the wheel makes an awful lot of sense! SedFlux is a ropes-and-rollers rabbit hole. Don't jump down it. 
We just need to keep this final bit of the triangle modelling very, very simple. It is the only way it will happen successfully I suspect
So my strong advice if to start with an algorithm that entrains sediment, perhaps with a control by water depth or wave energy if that is calculated, transports it in a unidirectional manner, and dumps it. As Xi and Burgess 2020 does. With that in place we could then make it more complex, but to make it more complex than that from the start means this triangle paper is likely to be published sometime in the 2040s ;-)

**Comment by HannoSpreeuw** [comment]
> SedFlux is a ropes-and-rollers rabbit hole.

So do I get you right that you have tried running it, is that what your experience is based on, or rather what you have inferred from the paper?

**Comment by HannoSpreeuw** [comment]
I guess you can infer the complexity from the title alone: _Sedflux 2.0: An advanced process-response model that generates three-dimensional stratigraphy_.

**Comment by EmiliaJarochowska** [comment]
I've ran it but from that it's not necessarily obvious how complex the algorithm is. You just get the result. But isn't the paper clear about the level of complexity?

**Comment by HannoSpreeuw** [comment]
Yes, that is clear to me now.

I guess the only way we could be using it is to compare it with the output of runs based on [Johan's proposal](https://github.com/MindTheGap-ERC/CarboKitten.jl/blob/73-onshore-transport/docs/src/onshore-transport.md).

That would require choosing parameter settings in Sedflux such that it is effectively a much simpler model i.e. closer to  [Johan's proposal](https://github.com/MindTheGap-ERC/CarboKitten.jl/blob/73-onshore-transport/docs/src/onshore-transport.md).
But this is just an idea, I did not investigate Sedflux options; it may not be possible to simplify Sedflux in this way.

**Comment by Burgesski** [comment]
> > SedFlux is a ropes-and-rollers rabbit hole.
> 
> So do I get you right that you have tried running it, is that what your experience is based on, or rather what you have inferred from the paper?

I worked with SedFlux a little back in the late 2000s and early 2010s as part of my broader involvement with the CSDMS group. I had a few longish discussions with the CSDMS software engineers and geoscientists working with SedFlex about the model data and object structure, for example, which is elegant in lots of ways but I suspect would be non-trivial to integrate with CarboKitten. Back then I wrote some very simple code for them to incorporate carbonate functionality in SedFlux but I don't think they ever really pursued that.

Anyway, my "please keep it simple" pleas are mostly derived from this experience; I've been making stratigraphic models for over 30 years (old, so old...), so I've seen a fair few rabbit holes, or, as Roy Batty put it in Blade Runner "I've seen things you people wouldn't believe" ;-)

**Comment by HannoSpreeuw** [comment]
In the meantime I have been consulting the great @jhidding and it turned out he had already compiled [a notebook](https://github.com/MindTheGap-ERC/CarboKitten.jl/blob/73-onshore-transport/notebooks/onshore.jl) for runs based on [his model](https://github.com/MindTheGap-ERC/CarboKitten.jl/blob/73-onshore-transport/docs/src/onshore-transport.md).

He demonstrated its use:
The velocity is a linear function of depth with a sign switch below some depth, modelled [here](https://github.com/MindTheGap-ERC/CarboKitten.jl/blob/7be52a6d3bd286d964eab9b905743f74995fe402/notebooks/onshore.jl#L46).
And you can play with the eustasy frequency [here](https://github.com/MindTheGap-ERC/CarboKitten.jl/blob/7be52a6d3bd286d964eab9b905743f74995fe402/notebooks/onshore.jl#L75).

I am currently playing with the [onshore velocity](https://github.com/MindTheGap-ERC/CarboKitten.jl/blob/7be52a6d3bd286d964eab9b905743f74995fe402/notebooks/onshore.jl#L57) to see if I can generate rhythmic cycles.
Perhaps we need a different kind of plot to visualize those cycles.

**Comment by jhidding** [comment]
I seem to remember we need to switch off sealevel variation for this one, and also disable the CA. I can help with the latter. Probably the simplest case should have two facies types?

**Comment by EmiliaJarochowska** [comment]
In the paper by Xi and @Burgesski it works both with CA on and off. And 3 facies. So maybe let's not worry about that? It is a matter of balance between transport and subsidence so maybe one could bruteforce loop through a bunch of combinations of subsidence and velocity to screen for any rhythmic behavior?

@HannoSpreeuw would the stratigraphic column be more useful than the slice plot? I cannot check today unfortunately.

**Comment by Burgesski** [comment]
Testing for cyclicity in the strata from the model is always a bit tricky. Simplest approach is to plot a transport-direction-parallel cross-section and chronostrat diagram, colour coded by depositional water depth, and watch out for stripes e.g. 

![Image](https://github.com/user-attachments/assets/e62247b7-021a-4204-a77d-b92bbf43e652) chronostrat diagram examples from Burgess and Wright (2003) - blue green is shallow water, yellow-red is intra and supratidal shoreline strata, white is hiatus

no CA would likely look more like that Case 2, with CA perhaps more irregular like Case 3

sometimes, to get started, visual inspection rocks! ;-)

cheers, P.

**Comment by Burgesski** [comment]
I think equation 1 in Xi and Burgess (2022) gives the velocity profiles, although in a slightly odd form - if you calculate the right hand side of the that equation with a range of water depths 0-20m you will get the velocity profile, obviously. The profile we used on those older papers is shown in Fig 1 from Burgess et al. (2001) 

![Image](https://github.com/user-attachments/assets/f370016d-7a82-4b01-82a0-0476ecd0ef4a)

So something like this would also work I think. Basically, I suspect that so long as there is subsidence/relative sealevel rise combined with  marine subtidal unidirectional transport that dumps transported sediment in front of any subaerial barrier, the effect will occur in some form

But of course, as we know from a certain other example, I could be wrong ;-)

**Comment by Burgesski** [comment]
Ah, fond memories -  "All three authors would like to thank G. Lucas for many helpful philosophical insights," is still one of my favourite acknowledgements. Insightful he is, mmmh ;-)

**Comment by EmiliaJarochowska** [comment]
@Burgesski we are almost there but still confused between wave velocity and sediment transport. Eq. 1 in Burgess et al. (2001) gives the velocity of the water particle on the **water surface** as a function of depth under the wave, but not the velocity profile in the water column. 

And then sediment transport will have a different profile since it has a different direction in the upper and lower part of the water column between sediment surface and the seafloor (shoreward in the upper part, basinward in the lower part). This seems to be reflected in the figure you pasted, where "sediment transport" is max at depth of 10 and decreases symmetrically upwards and downwards but how is it derived from water velocity?

**Comment by Burgesski** [comment]
All the various forms of this relationship Burgess et al have used have been a bit unsatisfactory in my opinion. But, that said, I'm not sure it really matters, because the more I think about  it, the more I think it is probably mostly meaningless to model waves and water velocity explicitly in any way when the model time step is 100 or 1000 years, or perhaps even longer. 

We need something that is time averaged, but we have zero data on time averaged marine process of this type over the required time scale. So in the end we are really just guessing. And best guess is the simplest guess, or so Mr Occam and others would argue. 

Hence why I keep banging on about the simplest possible model, just a simple constant rate landward flux. Start with that, add details like some sort of time averaged depth-rate relationship until something interesting happens. Then you know what needs to be assumed for that interesting effect to occur, and you can judge how time-averaged realistic that might or might not be...

Does that make sense? Or am I just lost in Model Land? ;-)

**Comment by Burgesski** [comment]
Two things strike me in response:
1. If one knows shear stress due to a combination of water flow and topographic slope, then it is relatively straight forward to determine probability of entrainment as bedload or suspended load for any grainsize, as Warrlich et al did. I've never understood why calculating sediment flux on this basis is a problem?
2. We "don't want sediment accumulating as if it was transported upwards on a steep slope as a result of onshore wave transport." Actually, why not? Surely this is exactly what waves and tides and in fact even wind-induced unidirectional currents on real platform tops do? And time averaged (e.g. 1ky time step) this probably includes big enough waves to transport up much steeper slopes than occur across the scale of a grid cell in any model we will run

So, I'm not sure any of these things are sufficiently good reason to not just keep it very simple, as previously suggested?

but, if you really do want a specific representation of waves in the model, the sine function looks OK to me, subject to good testing behaviour obvs

cheers, P.

**Comment by HannoSpreeuw** [comment]
I don't yet see how [Stokes drift](https://en.wikipedia.org/wiki/Stokes_drift) should be combined with [Airy wave theory](https://en.wikipedia.org/wiki/Airy_wave_theory), as Stokes drift essentially emerges from Airy wave theory, as far as I understood.
Which means that this description:

![Image](https://github.com/user-attachments/assets/45f2cb1c-ce2c-46f1-ab2b-de7a4ff2da8c)

would be all we need for deep water waves.

I guess the more general formula - for basins of arbitrary depth - is listed on the [Airy wave theory Wikipedia page](https://en.wikipedia.org/wiki/Airy_wave_theory) as

![Image](https://github.com/user-attachments/assets/82db41c3-7674-4ed3-bda6-0d5f35706c83)

This is the Stokes drift defined as the particle drift after one wave cycle divided by the period.
Does this converge to the formula above for deep-water waves when $h$ is very large? 
Let me check with pen and paper.

**Comment by HannoSpreeuw** [comment]
The latter expression indeed converges to
$\sigma \cdot k \cdot a² e^{2\cdot k \cdot z} = \sigma \cdot k \cdot a² e^{\frac{4\cdot \pi \cdot z}{\lambda}} $
for deep water, which is equal to the former expression if $\omega == \sigma$.

**Comment by jhidding** [comment]
Again, I think you're getting bogged down in details. We were a bit flummoxed by the tanh profile before, as our intuition told us that for very deep waters the transport should converge to zero, not increase to some maximum. We found the correct principles and the generic shape that matches our intuition. Then it is time to start running the model. Just use any reasonable profile and see how the model behaves.

**Comment by HannoSpreeuw** [comment]
Allright then.

But I am still puzzled about why you return the [derivative wrt depth of the velocity profile](https://github.com/MindTheGap-ERC/CarboKitten.jl/blob/11b74f1436c7bf3e0af504557404e8c314ee9781/notebooks/onshore.jl#L47).

Shouldn't we use the velocity decay over horizontal distance - similar to water moving from a fast-flowing river to a slow delta - instead?

**Comment by EmiliaJarochowska** [comment]
We are implicitly making an assumption that our time-averaged waves are triggered (again, on average) all the time and always in onshore direction. So they are not generated at a point or line, but uniformly everywhere in the basin. So our time- and space-averaged wind acts everywhere. That is different from river to delta transition where the same water flux flows from a smaller to a larger diameter.

**Comment by EmiliaJarochowska** [comment]
Yeah maybe that needs some reflection 🤔 but let's treat these two models as separate modules

**Comment by jhidding** [comment]
The assumption in the current implementation is that the added sediment flux is not directly dependent on space, only on water depth. So we have an additional term $v(w)$ in the flux $q$. Since $h$ (the topography) varies as a function of space, the added flux also implies spatial derivatives through the chain rule. Here, also the derivative $v'(w)$ pops up, which could be interpreted as the velocity shear. It could very well be that there is a sign error in my prototype implementation of a first velocity profile, which was just a linear relation. In most of our models geometry is specified such that onshore transport is in the negative $x$ direction. By all means, throw that in the trash can.

**Comment by HannoSpreeuw** [comment]
Allright.

When you look at [your model](https://github.com/MindTheGap-ERC/CarboKitten.jl/blob/73-onshore-transport/docs/src/onshore-transport.md#onshore-transport), you have, as the final equation:

$$\partial_t \eta_f = \nu_f P_f \nabla^2 \eta + (\nu_f \nabla P_f - P_f w_f') \cdot \nabla \eta - \nabla P_f \cdot w_f + P_f.$$,

i.e. $w_f$ and $w_f'$ both get a minus sign, but in [the implementation](https://github.com/MindTheGap-ERC/CarboKitten.jl/blob/73-onshore-transport/src/Components/ActiveLayerOnshore.jl#L17)
only $ss$ - the derivative of the onshore velocity wrt depth - [gets a minus sign](https://github.com/MindTheGap-ERC/CarboKitten.jl/blob/11b74f1436c7bf3e0af504557404e8c314ee9781/src/Components/ActiveLayerOnshore.jl#L20):
```
adv = -((x[3, 2][1] - x[1, 2][1]) / (2Δx) * (d * (x[3, 2][2] - x[1, 2][2]) / (2Δx) - P * ss[1] * Δt) +
        (x[2, 3][1] - x[2, 1][1]) / (2Δx) * (d * (x[2, 3][2] - x[2, 1][2]) / (2Δx) - P * ss[2] * Δt))
```

while $sv$ - the onshore velocity, [does not](https://github.com/MindTheGap-ERC/CarboKitten.jl/blob/11b74f1436c7bf3e0af504557404e8c314ee9781/src/Components/ActiveLayerOnshore.jl#L26):
```
    prd = sv[1] * (x[3, 2][2] - x[1, 2][2]) * Δt / (2Δx) + sv[2] * (x[2, 3][2] - x[2, 1][2]) * Δt / (2Δx) + x[2, 2][2]
```

So [the way you set it up](https://github.com/MindTheGap-ERC/CarboKitten.jl/blob/11b74f1436c7bf3e0af504557404e8c314ee9781/notebooks/onshore.jl#L46C1-L47C50), by propagating the onshore velocity with a minus and its derivative wrt depth without:
```
			[(-max(0.0u"m/yr", b * (a - w) / a), 0.0u"m/yr"),
			 ((w <= a ? -b/a : 0.0u"yr^-1"), 0.0u"yr^-1")]
```

works out just fine.

**Comment by EmiliaJarochowska** [comment]
Thanks for finding this, @jhidding . Should this be a new issue?

**Comment by EmiliaJarochowska** [comment]
And does it help you, @HannoSpreeuw ?

**Comment by HannoSpreeuw** [comment]
Indeed. I can fix the code and then start runs for different parameters.

**Comment by HannoSpreeuw** [comment]
@EmiliaJarochowska After the recent pushes to the `73-branch` and my discussion with @jhidding this morning, we would like to have your say on the next step(s) forward.

The current situation is that we have implemented `Johan's velocity profile` in the [notebook](https://github.com/MindTheGap-ERC/CarboKitten.jl/blob/73-onshore-transport/notebooks/onshore.jl). It integrates without derailing for the current set of two parameters - `v_max==10.0u"m/yr"` and `max_depth==20.0u"m"`. 
We do not find autocyclic patters with these settings.
(Note  that `max_depth` is the depth at which the onshore velocity is maximal.)  
If you change `v_max` to, e.g., `200.0u"m/yr"`, the integration will derail.

Some options to proceed:

- [ ] Explore parameter space for `v_max` and `max_depth` to find autocyclic patterns, using the current Eulerian scheme.
- [ ] Explore parameter space for other parameters, like diffusion and extinction coefficients.
- [ ] Do not do so yet, first modify the integrations scheme to include the `upwind solution` for better stability.
- [ ] Do not do so yet, first add an integration engine like [DifferentialEquations.jl](https://docs.sciml.ai/DiffEqDocs/stable/) to include implicit schemes.
 
What is your pick for this?

**Comment by Burgesski** [comment]
Hm, I'm not entirely surprised. Before we start exploring the parameter space, perhaps a simpler check on what this model is doing would be a good idea. So, in this formulation, two things to check:
1. Is there a flux of sediment across the platform top 
2. Does the flux get deposited at the point where grid cell elevation is above sea-level?
If these two conditions are met. autocycles should form, or at least a shoreline should prograde. 

Out of interest, see Demicco 98 (attached, hopefully) and also GOLDHAMMER, R.K., LEHMANN, P.J., AND DUNN, P.A., 1993, The origin of high-frequency platform carbonate cycles and third-order sequences (Lower Ordovician El Paso Group, West Texas): Constraints from outcrop data and stratigraphic modeling: Journal of Sedimentary Petrology, v. 63, p. 318–359 for previous models that showed this same basic autocyclic process, just so you know I'm not making up crazy autoycle effects, or at least if I am, other people did it first ;-)

[demicco 98 cyclopath.pdf](https://github.com/user-attachments/files/19167907/demicco.98.cyclopath.pdf)

**Comment by jhidding** [comment]
Hanno, you forget some of the most important parts:

- Run 1d tests similar to the ActiveLayer tests, with pure advection, pure diffusion and mixed cases.
- Also have a case with the velocity constant and the shear zero.
- disable CA
- constant sea level
